### PR TITLE
aws_ec2_transit_gateway_route: cidr match is unreliable

### LINKED ISSUE #14845: aws_ec2_transit_gateway_route does not properly compare compressed/expanded ipv6 prefixes
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

Terraform v0.12.29
+ provider.aws v3.3.0

### Affected Resource(s)

* aws_ec2_transit_gateway_route

### Terraform Configuration Files
Failure:
```hcl
provider "aws" {
  region    = "us-east-1"
  version   = "~> 3.3"
}
resource "aws_ec2_transit_gateway_route" "test" {
  destination_cidr_block         = "2001:db8::/56"
  transit_gateway_attachment_id  = "tgw-attach-04ea4806ba7537e56"
  transit_gateway_route_table_id = "tgw-rtb-0c670a1a40e75389d"
}
```
Workaround:
```hcl
provider "aws" {
  region    = "us-east-1"
  version   = "~> 3.3"
}
resource "aws_ec2_transit_gateway_route" "test" {
  destination_cidr_block         = "2001:db8:0:0:0:0:0:0/56"
  transit_gateway_attachment_id  = "tgw-attach-04ea4806ba7537e56"
  transit_gateway_route_table_id = "tgw-rtb-0c670a1a40e75389d"
}
```

### Debug Output

Full debug omitted. PR incoming.

```
Error: Provider produced inconsistent result after apply

When applying changes to
aws_ec2_transit_gateway_route.test,
provider "registry.terraform.io/-/aws" produced an unexpected new value for
was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

### Expected Behavior

IPv6 CIDR should match route whether compressed or uncompressed format.

### Actual Behavior

Route only matches if uncompressed (:0:0:0:0:0:0:0...).

### Steps to Reproduce

Apply the sample terraform module.

### References

* #14846

### PR DESCRIPTION
**Pull request by thekev**
* aws api returns expanded prefix, but hcl may contain compressed prefix
* use cidrBlocksEqual() instead of string comparison to find route
* normalize route destination value to RFC1924 compressed

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #14845

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
Improve reliability of route matching for transit gateway routes.
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSEc2TransitGatewayRoute_basic_ipv6' ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 3 -run=TestAccAWSEc2TransitGatewayRoute_basic_ipv6 -timeout 120m
=== RUN   TestAccAWSEc2TransitGatewayRoute_basic_ipv6
=== PAUSE TestAccAWSEc2TransitGatewayRoute_basic_ipv6
=== CONT  TestAccAWSEc2TransitGatewayRoute_basic_ipv6
--- PASS: TestAccAWSEc2TransitGatewayRoute_basic_ipv6 (327.71s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	329.745s
$ make testacc TESTARGS='-run=Test_canonicalCidrBlock' ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 3 -run=Test_canonicalCidrBlock -timeout 120m
=== RUN   Test_canonicalCidrBlock
--- PASS: Test_canonicalCidrBlock (0.00s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1.823s
```

### CODE REVIEW

File: aws/ec2_transit_gateway.go [hunk 1]
```diff
@@ -109,8 +110,11 @@ func ec2DescribeTransitGatewayRoute(conn *ec2.EC2, transitGatewayRouteTableID, d
 		if route == nil {
 			continue
 		}
-
-		if aws.StringValue(route.DestinationCidrBlock) == destination {
+		if cidrBlocksEqual(aws.StringValue(route.DestinationCidrBlock), destination) {
+			// normalize IPv6 to RFC1924 compressed
+			_, ipnet, _ := net.ParseCIDR(aws.StringValue(route.DestinationCidrBlock))
+			cidrString := ipnet.String()
+			route.DestinationCidrBlock = &cidrString
```
Comment (ewbankkit) [inline aws/ec2_transit_gateway.go:117]: I hit a similar issue over in #14543 and ended up adding a new function (I put it in `validators.go` to keep it close to the other CIDR functions):

```go
// canonicalCidrBlock returns the canonical representation of a CIDR block.
// This function is especially useful for hash functions for sets which include IPv6 CIDR blocks.
func canonicalCidrBlock(cidr string) string {
	_, ipnet, err := net.ParseCIDR(cidr)
	if err != nil {
		return cidr
	}

	return ipnet.String()
}
```

Comment (thekev) [reply aws/ec2_transit_gateway.go:117]: I like this. Looks like your PR isn’t ready yet? Mind if I steal this snippet? It’s cleaner than what I did.

Comment (ewbankkit) [reply aws/ec2_transit_gateway.go:117]: Sure, please do.

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSEc2TransitGatewayRoute_basic_ipv6'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSEc2TransitGatewayRoute_basic_ipv6 -timeout 120m
=== RUN   TestAccAWSEc2TransitGatewayRoute_basic_ipv6
=== PAUSE TestAccAWSEc2TransitGatewayRoute_basic_ipv6
=== CONT  TestAccAWSEc2TransitGatewayRoute_basic_ipv6
--- PASS: TestAccAWSEc2TransitGatewayRoute_basic_ipv6 (306.79s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	306.826s
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSEc2TransitGatewayRoute_basic$$'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSEc2TransitGatewayRoute_basic$ -timeout 120m
=== RUN   TestAccAWSEc2TransitGatewayRoute_basic
=== PAUSE TestAccAWSEc2TransitGatewayRoute_basic
=== CONT  TestAccAWSEc2TransitGatewayRoute_basic
--- PASS: TestAccAWSEc2TransitGatewayRoute_basic (304.09s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	304.134s
```

File: aws/ec2_transit_gateway.go [hunk 2]
```diff
@@ -109,8 +109,9 @@ func ec2DescribeTransitGatewayRoute(conn *ec2.EC2, transitGatewayRouteTableID, d
 		if route == nil {
 			continue
 		}
-
-		if aws.StringValue(route.DestinationCidrBlock) == destination {
+		if cidrBlocksEqual(aws.StringValue(route.DestinationCidrBlock), destination) {
+			cidrString := canonicalCidrBlock(aws.StringValue(route.DestinationCidrBlock))
+			route.DestinationCidrBlock = &cidrString
```
Comment (ewbankkit) [inline aws/ec2_transit_gateway.go:114]: Should use `aws.String(cidrString)` here instead of taking raw address.

Comment (ewbankkit) [inline aws/ec2_transit_gateway.go:114]:
```suggestion
			route.DestinationCidrBlock = aws.String(cidrString)
```

**Review by DrFaust92 — APPROVED** [review]
LGTM

```
--- PASS: TestAccAWSEc2TransitGatewayRoute_basic_ipv6 (382.25s)
```

**Review by gdavison — APPROVED** [review]
LGTM 🚀 

--- PASS: TestCanonicalCidrBlock (0.00s)
--- PASS: TestAccAWSEc2TransitGatewayRouteTable_disappears_TransitGateway (254.06s)
--- PASS: TestAccAWSEc2TransitGatewayRouteTable_disappears (308.93s)
--- PASS: TestAccAWSEc2TransitGatewayRoute_disappears (369.18s)
--- PASS: TestAccAWSEc2TransitGatewayRoute_disappears_TransitGatewayAttachment (371.33s)
--- PASS: TestAccAWSEc2TransitGatewayRouteTable_basic (372.06s)
--- PASS: TestAccAWSEc2TransitGatewayRouteTableDataSource_ID (371.42s)
--- PASS: TestAccAWSEc2TransitGatewayRouteTablePropagation_basic (372.35s)
--- PASS: TestAccAWSEc2TransitGatewayRouteTableDataSource_Filter (372.46s)
--- PASS: TestAccAWSEc2TransitGatewayRouteTable_Tags (372.87s)
--- PASS: TestAccAWSEc2TransitGatewayRoute_basic (431.73s)
--- PASS: TestAccAWSEc2TransitGatewayRoute_blackhole (431.61s)
--- PASS: TestAccAWSEc2TransitGatewayRoute_basic_ipv6 (433.33s)
--- PASS: TestAccAWSEc2TransitGatewayRouteTableAssociation_basic (492.15s)

**Pull request merged**